packages/app/src/systems/Ecosystem/machines/ecosystemMachine.tsx

```diff
@@ -0,0 +1,235 @@
+          },
+          CLEAR_FILTER: {
+            actions: ['clearFilter'],
+          },
+        },
+      },
+      fetching: {
+        tags: ['isLoading'],
+        invoke: {
+          src: 'fetchProjectsAndTags',
+          onDone: {
+            target: 'idle',
+            actions: ['assignProjectsAndTags', 'clearSearchAndFilter'],
+          },
+        },
+      },
+      filtering: {
+        entry: ['clearSearch'],
+        tags: ['isLoading'],
+        invoke: {
+          src: 'filterProjects',
+          data: {
+            input: (_: MachineContext, ev: EcosystemMachineEvents) => ({
+              ...ev.input,
+              projects: _.projects,
+            }),
+          },
+          onDone: {
+            target: 'idle',
+            actions: ['assignFilteredProjects'],
+          },
+        },
+      },
+      searching: {
+        entry: ['clearFilter'],
+        tags: ['isLoading'],
+        invoke: {
+          src: 'searchProjects',
+          data: {
+            input: (_: MachineContext, ev: EcosystemMachineEvents) => ({
+              ...ev.input,
+              projects: _.projects,
+            }),
+          },
+          onDone: {
+            target: 'idle',
+            actions: ['assignFilteredProjects'],
+          },
+        },
+      },
+    },
+  },
+  {
+    services: {
+      fetchProjectsAndTags: FetchMachine.create<
+        null,
+        { projects: Project[]; tags: string[] }
+      >({
+        showError: true,
+        maxAttempts: 1,
+        async fetch() {
+          const projects = SAMPLE_PROJECTS as Project[];
+          const tags = new Set<string>('');
+          projects.map((project) => project.tags.map((tag) => tags.add(tag)));
+          return {
+            projects,
+            tags: Array.from(tags),
+          };
+        },
+      }),
+      filterProjects: FetchMachine.create<EcosystemInputs['filter'], Project[]>(
+        {
+          showError: true,
+          maxAttempts: 1,
+          async fetch({ input }) {
+            if (!input?.projects) throw new Error('Invalid projects');
+            if (!input?.tag) return input.projects;
+            const { tag, projects } = input;
+            const filteredProjects = projects.filter((project) => {
+              return project.tags.includes(tag);
+            });
+            return filteredProjects ?? [];
+          },
+        }
+      ),
+      searchProjects: FetchMachine.create<EcosystemInputs['search'], Project[]>(
+        {
+          showError: true,
+          maxAttempts: 1,
+          async fetch({ input }) {
+            if (!input?.projects) throw new Error('Invalid projects');
+            if (!input?.query) return input.projects;
+            const { query, projects } = input;
+            const filteredProjects = projects.filter((project) => {
+              return project.name.toLowerCase().includes(query.toLowerCase());
+            });
+            return filteredProjects ?? [];
+          },
```

Comment:
it's not fetching, this would be more suitable to be a selector in the hook?

Reply:
hmm